File: python-sdk/tests/test_audio_verbs.py
"""
Audio verbs + retrieval helpers (voice-call-io task 08, AC8/AC11).

Verifies the SDK exposes ``inject_audio`` / ``start_audio_capture`` /
``stop_audio_capture`` mirroring the recording-verb precedent, surfaces the
wire receipt's ``audio_after_hash`` / ``audio_stop_reason``, and fetches the
captured WAV bytes via ``content.get`` (``fetch_audio_capture`` /
``save_audio_capture``) without touching the internal ContentStore API.
"""

from __future__ import annotations

import json
import logging
from typing import Any

import loom
from loom.types import Receipt

AUDIO_HASH = "9f" * 32
WAV_BYTES = b"RIFF\x24\x00\x00\x00WAVEfmt " + bytes(28)


def _decode_payload(params: dict) -> dict:
    return json.loads(bytes(params["action"]["payload"]).decode("utf-8"))


def _stop_receipt(stop_reason: str) -> dict:
    return {
        "action_hash": "a" * 64,
        "outcome_hash": "b" * 64,
        "emitted_at_ms": 1_730_000_000_000,
        "audio_after_hash": AUDIO_HASH,
        "audio_stop_reason": stop_reason,
    }


def _register_audio(daemon: Any, stop_reason: str = "explicit") -> dict[str, Any]:
    state: dict[str, Any] = {"inject": None, "start": None, "stop_calls": 0}

    def _inject(params: dict) -> dict:
        state["inject"] = _decode_payload(params)
        return {"action_hash": "a" * 64, "outcome_hash": "b" * 64, "emitted_at_ms": 1}

    def _start(params: dict) -> dict:
        state["start"] = _decode_payload(params)
        return {"action_hash": "a" * 64, "outcome_hash": "b" * 64, "emitted_at_ms": 1}

    def _stop(params: dict) -> dict:
        state["stop_calls"] += 1
        return _stop_receipt(stop_reason)

    daemon.register_handler("action.web.inject_audio", _inject)
    daemon.register_handler("action.web.start_audio_capture", _start)
    daemon.register_handler("action.web.stop_audio_capture", _stop)
    return state


def _register_content_get(daemon: Any) -> dict[str, Any]:
    state: dict[str, Any] = {"artifact_ref": None}

    def _content_get(params: dict) -> dict:
        state["artifact_ref"] = params["artifact_ref"]
        return {
            "artifact_ref": params["artifact_ref"],
            "data_hex": WAV_BYTES.hex(),
            "size_bytes": len(WAV_BYTES),
        }

    daemon.register_handler("content.get", _content_get)
    return state


def test_inject_audio_threads_payload(daemon):
    state = _register_audio(daemon)
    with loom.Session.create(socket_path=str(daemon.socket_path), token=daemon.token) as s:
        receipt = s.inject_audio(blob_ref=AUDIO_HASH, await_playout=True)

    assert state["inject"]["blob_ref"] == AUDIO_HASH
    assert state["inject"]["await_playout"] is True
    assert "audio_b64" not in state["inject"]
    assert isinstance(receipt, Receipt)


def test_start_audio_capture_threads_caps(daemon):
    state = _register_audio(daemon)
    with loom.Session.create(socket_path=str(daemon.socket_path), token=daemon.token) as s:
        s.start_audio_capture(max_duration_ms=30_000, max_bytes=1_000_000)

    assert state["start"]["max_duration_ms"] == 30_000
    assert state["start"]["max_bytes"] == 1_000_000


def test_stop_audio_capture_surfaces_audio_fields(daemon):
    _register_audio(daemon)
    with loom.Session.create(socket_path=str(daemon.socket_path), token=daemon.token) as s:
        s.start_audio_capture()
        receipt = s.stop_audio_capture()

    assert receipt.audio_after_hash == AUDIO_HASH
    assert receipt.audio_stop_reason == "explicit"


def test_stop_audio_capture_warns_on_cap_truncation(daemon, caplog):
    for reason in ("byte_cap", "duration_cap"):
        caplog.clear()
        _register_audio(daemon, stop_reason=reason)
        with loom.Session.create(socket_path=str(daemon.socket_path), token=daemon.token) as s:
            with caplog.at_level(logging.WARNING, logger="loom"):
                receipt = s.stop_audio_capture()

        assert receipt.audio_stop_reason == reason
        warn_messages = [rec.message for rec in caplog.records]
        assert any(reason in m and "truncated" in m for m in warn_messages), warn_messages


def test_stop_audio_capture_does_not_warn_on_explicit(daemon, caplog):
    _register_audio(daemon, stop_reason="explicit")
    with loom.Session.create(socket_path=str(daemon.socket_path), token=daemon.token) as s:
        with caplog.at_level(logging.WARNING, logger="loom"):
            s.stop_audio_capture()

    assert not caplog.records


def test_fetch_audio_capture_returns_bytes(daemon):
    _register_audio(daemon)
    content_state = _register_content_get(daemon)
    with loom.Session.create(socket_path=str(daemon.socket_path), token=daemon.token) as s:
        receipt = s.stop_audio_capture()
        data = s.fetch_audio_capture(receipt)

    assert data == WAV_BYTES
    assert content_state["artifact_ref"] == AUDIO_HASH


def test_save_audio_capture_writes_file(daemon, tmp_path):
    _register_audio(daemon)
    _register_content_get(daemon)
    out = tmp_path / "answer.wav"
    with loom.Session.create(socket_path=str(daemon.socket_path), token=daemon.token) as s:
        receipt = s.stop_audio_capture()
        s.save_audio_capture(receipt, str(out))

    assert out.read_bytes() == WAV_BYTES


def test_fetch_audio_capture_without_hash_raises(daemon):
    with loom.Session.create(socket_path=str(daemon.socket_path), token=daemon.token) as s:
        bare = Receipt._from_dict({"action_hash": "a" * 64, "outcome_hash": "b" * 64})
        try:
            s.fetch_audio_capture(bare)
        except ValueError as exc:
            assert "audio_after_hash" in str(exc)
        else:
            raise AssertionError("expected ValueError for receipt without audio_after_hash")

File: typescript-sdk/tests/test_audio_verbs.ts
/**
 * Audio verbs + retrieval helpers (voice-call-io task 08, AC8/AC11).
 *
 * Verifies the SDK exposes `injectAudio` / `startAudioCapture` /
 * `stopAudioCapture`, surfaces the wire receipt's `audio_after_hash` /
 * `audio_stop_reason` as `audioAfterHash` / `audioStopReason`, warns
 * loudly on cap truncation, and fetches the captured WAV bytes via
 * `content.get` (`fetchAudioCapture` / `saveAudioCapture`).
 */
import { test, describe, beforeEach, afterEach } from "node:test";
import assert from "node:assert/strict";
import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import { Session, type Receipt } from "../src/index.js";
import { MockDaemon } from "./helpers/mock_daemon.js";

const AUDIO_HASH = "9f".repeat(32);
const WAV_BYTES = Buffer.concat([Buffer.from("RIFF\x24\x00\x00\x00WAVEfmt "), Buffer.alloc(28)]);

/** Decode the JSON action payload the SDK packs into `action.payload` bytes. */
function decodePayload(params: Record<string, unknown>): Record<string, unknown> {
  const action = params["action"] as Record<string, unknown>;
  const bytes = action["payload"] as number[];
  return JSON.parse(Buffer.from(bytes).toString("utf8")) as Record<string, unknown>;
}

function stopReceipt(stopReason: string): Record<string, unknown> {
  return {
    action_hash: "a".repeat(64),
    outcome_hash: "b".repeat(64),
    emitted_at_ms: 1_730_000_000_000,
    audio_after_hash: AUDIO_HASH,
    audio_stop_reason: stopReason,
  };
}

describe("audio verbs + retrieval helpers", () => {
  let daemon: MockDaemon;
  let injectPayload: Record<string, unknown> = {};
  let startPayload: Record<string, unknown> = {};
  let stopReason = "explicit";
  let contentGetRef: string | undefined;
  let warnings: string[] = [];
  const origWarn = console.warn;

  beforeEach(async () => {
    daemon = new MockDaemon();
    daemon.registerHandler("action.web.inject_audio", (params) => {
      injectPayload = decodePayload(params);
      return { action_hash: "a".repeat(64), outcome_hash: "b".repeat(64), emitted_at_ms: 1 };
    });
    daemon.registerHandler("action.web.start_audio_capture", (params) => {
      startPayload = decodePayload(params);
      return { action_hash: "a".repeat(64), outcome_hash: "b".repeat(64), emitted_at_ms: 1 };
    });
    daemon.registerHandler("action.web.stop_audio_capture", () => stopReceipt(stopReason));
    daemon.registerHandler("content.get", (params) => {
      contentGetRef = params["artifact_ref"] as string;
      return {
        artifact_ref: contentGetRef,
        data_hex: WAV_BYTES.toString("hex"),
        size_bytes: WAV_BYTES.length,
      };
    });
    await daemon.start();
    stopReason = "explicit";
    contentGetRef = undefined;
    warnings = [];
    console.warn = (...args: unknown[]) => {
      warnings.push(args.map(String).join(" "));
    };
  });

  afterEach(async () => {
    console.warn = origWarn;
    await daemon.stop();
  });

  test("injectAudio threads blob_ref/await_playout into the payload", async () => {
    const s = await Session.create({ socketPath: daemon.socketPath, token: daemon.token });
    const receipt = await s.injectAudio({ blobRef: AUDIO_HASH, awaitPlayout: true });
    await s.close();

    assert.equal(injectPayload["blob_ref"], AUDIO_HASH);
    assert.equal(injectPayload["await_playout"], true);
    assert.equal("audio_b64" in injectPayload, false);
    assert.equal(receipt.ok, true);
  });

  test("startAudioCapture threads caps into the payload", async () => {
    const s = await Session.create({ socketPath: daemon.socketPath, token: daemon.token });
    await s.startAudioCapture({ maxDurationMs: 30_000, maxBytes: 1_000_000 });
    await s.close();

    assert.equal(startPayload["max_duration_ms"], 30_000);
    assert.equal(startPayload["max_bytes"], 1_000_000);
  });

  test("stopAudioCapture surfaces audioAfterHash + audioStopReason", async () => {
    const s = await Session.create({ socketPath: daemon.socketPath, token: daemon.token });
    const receipt = await s.stopAudioCapture();
    await s.close();

    assert.equal(receipt.audioAfterHash, AUDIO_HASH);
    assert.equal(receipt.audioStopReason, "explicit");
    assert.equal(warnings.length, 0, `explicit stop must not warn: ${warnings}`);
  });

  for (const reason of ["byte_cap", "duration_cap"]) {
    test(`stopAudioCapture warns loudly on ${reason} truncation`, async () => {
      stopReason = reason;
      const s = await Session.create({ socketPath: daemon.socketPath, token: daemon.token });
      const receipt = await s.stopAudioCapture();
      await s.close();

      assert.equal(receipt.audioStopReason, reason);
      assert.equal(warnings.length, 1);
      assert.ok(warnings[0].includes(reason), `warning must name the reason: ${warnings[0]}`);
      assert.ok(warnings[0].includes("truncated"), `warning must say truncated: ${warnings[0]}`);
    });
  }

  test("fetchAudioCapture returns the WAV bytes via content.get", async () => {
    const s = await Session.create({ socketPath: daemon.socketPath, token: daemon.token });
    const receipt = await s.stopAudioCapture();
    const bytes = await s.fetchAudioCapture(receipt);
    await s.close();

    assert.deepEqual(Buffer.from(bytes), WAV_BYTES);
    assert.equal(contentGetRef, AUDIO_HASH);
  });

  test("saveAudioCapture writes the WAV to disk", async () => {
    const dir = fs.mkdtempSync(path.join(os.tmpdir(), "loom-audio-"));
    const out = path.join(dir, "answer.wav");
    try {
      const s = await Session.create({ socketPath: daemon.socketPath, token: daemon.token });
      const receipt = await s.stopAudioCapture();
      await s.saveAudioCapture(receipt, out);
      await s.close();

      assert.deepEqual(fs.readFileSync(out), WAV_BYTES);
    } finally {
      fs.rmSync(dir, { recursive: true, force: true });
    }
  });

  test("fetchAudioCapture throws on malformed data_hex instead of zero-filling", async () => {
    daemon.registerHandler("content.get", (params) => ({
      artifact_ref: params["artifact_ref"],
      data_hex: "zz-not-hex",
      size_bytes: 5,
    }));
    const s = await Session.create({ socketPath: daemon.socketPath, token: daemon.token });
    const receipt = await s.stopAudioCapture();
    await assert.rejects(
      () => s.fetchAudioCapture(receipt),
      (err: Error) => err.message.includes("malformed data_hex"),
    );
    await s.close();
  });

  test("fetchAudioCapture without audioAfterHash throws naming the field", async () => {
    const s = await Session.create({ socketPath: daemon.socketPath, token: daemon.token });
    const bare = { actionHash: "a".repeat(64), outcomeHash: "b".repeat(64), emittedAtMs: 1 } as Receipt;
    await assert.rejects(
      () => s.fetchAudioCapture(bare),
      (err: Error) => err.message.includes("audioAfterHash"),
    );
    await s.close();
  });
});
